### DISCUSSION TITLE
Consider adding ability to modify request headers in middleware for complex authentication schemes

### DISCUSSION THREAD

**Discussion by waza-ari** [fastapi/fastapi#6425] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Request
from jose import jwt, jwk


async def verify_pop_token(request: Request, call_next):

    # Verify and extract Proof of Possession Token if present
    if 'authorization' in request.headers and request.headers['authorization'].startswith('PoP '):

        # Fetch the PoP token from the header
        user_token = request.headers['authorization'].split(" ")[1]

        # Decode without verification to get the public key
        decoded_token = jwt.decode(user_token, "", algorithms=['RS256'], options={
            'verify_signature': False,
            'require': []
        })

        # Construct JWK key object
        public_key = jwk.construct(decoded_token['cnf']['jwk'])

        # Now try to verify the token with the key - this proves the sender has the private key
        verified_token = jwt.decode(user_token, public_key, algorithms=['RS256'])

        # TODO: !important Verify method, host and path according to RFC draft
        # https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop

        # Modify request to include the bearer token, such that following middlewares can work with the Bearer token
        bearer_token = verified_token['at']

        # Problem is here: I cannot insert the bearer token into the request

    response = await call_next(request)
    return response
```

### Description

DPoP or a Proof of Possession scheme is a coming up method to add security to common OAuth flows. Essentially, a SPA generates a key pair, claims a token from an identity provider and sends it to a backend. The main feature is that the access_token is not returned directly, but instead enveloped into an additional construct, containing the actual access_token, the public key and a signature created by the client with their private key. This way, the endpoint (FastAPI in this case) can verify that the sender is actually possessing the private key used to sign the JWT token. This is just a rough explanation to make my point, [details can be found in the IETF draft](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop).

The challenge now is that there are many middlewares helping to validate a JWT access_token, just one [example being fastapi-azure-auth](https://github.com/Intility/fastapi-azure-auth). All these libraries can verify the token with their respective backend, but they do expect the regular Bearer token.

Usually the authentication header would look like this:

```
Authorization: Bearer <access_token>
```

With Proof of Possession however, it looks like this:

```
Authorization: PoP <envelope<access_token>>
```

I now wanted to build a middleware that takes the PoP token and performs the necessary validations (extract key, verify signature, check method, host and so on), and then passes the actual Bearer token to whatever next middleware is there to validate the actual token.

With the current restriction of not being able to modify the request headers, I cannot modify the request though and are therefore unable to do the verification without also re-implementing the entire follow-up middleware expecting a Bearer token to be present.

### Wanted Solution

I'd like to be able to alter the request (specifically the headers in this case) to support such use cases.

### Wanted Code

```python
request.headers.set('Authorization', bearer_token)

# or

request.headers.update({'Authorization': bearer_token})
```

### Operating System

macOS

### Operating System Details

Doesn't matter

### FastAPI Version

0.68.1

### Python Version

3.9.7

### Additional Context

I'm not sure this is correct here in FastAPI or if this is a underlying Starlette issue - I somehow suspect the latter. Maybe there is a way to work around it though without the need of removing the DPoP envelope in the SPA, which defeats the purpose of it :)

**Comment by waza-ari** [comment, marked as answer]
I've been able to implement the idea by modifying the scope and re-creating the headers object. Something along those lines:

```
from starlette.datastructures import Headers

async def verify_pop_token(request: Request, call_next):

    # some magic happens here

    bearer_token = "Bearer %s" % verified_token['at']

    # Remove original authorization header
    request.scope['headers'] = [e for e in request.scope['headers'] if not e[0] == b'authorization'] 
    request.scope['headers'].append((b'authorization', bearer_token.encode()))

    # Re-create headers in case they've already been read and _headers is set
    # see https://github.com/encode/starlette/blob/d87c93eda4c546771fb51c60a21e8f245905f7f6/starlette/requests.py#L105-L109
    request._headers = Headers(scope=request.scope)

    response = await call_next(request)
```

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
